File: graphs/topo_sort.cpp
#include <bits/stdc++.h>
using namespace std;

void topo_sort(int v, vector<int> adj[])
{
    vector<int> indegree(v, 0);
    for (int i = 0; i < v; i++)
    {
        for (auto it : adj[i])
        {
            indegree[it]++;
        }
    }
    queue<int> q;
    for (int i = 0; i < v; i++)
    {
        if (indegree[i] == 0)
        {
            q.push(i);
        }
    }
    vector<int> topo;
    while (!q.empty())
    {
        int node = q.front();
        q.pop();
        topo.push_back(node);
        for (auto it : adj[node])
        {
            indegree[it]--;
            if (indegree[it] == 0)
            {
                q.push(it);
            }
        }
    }
    for (auto it : topo)
    {
        cout << it << " ";
    }
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
}

File: graphs/topo_sort.java
// this is used to check if the graph contains cycle or not, if the topo sort gives the same number of elements as the vertex that means no cycle else contains cycle

import java.util.*;

class Topo {

    public int[] topoSort(int v, List<List<Integer>> adj) {
        int[] indegree = new int[v];
        for (int i = 0; i < v; i++) {
            for (int it : adj.get(i)) {
                indegree[it]++;
            }
        }
        int[] topoSort = new int[v];
        int index = 0;
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < v; i++) {
            if (indegree[i] == 0) {
                q.offer(i);
            }
        }
        while (!q.isEmpty()) {
            int node = q.poll();
            topoSort[index++] = node;
            for (int neighbour : adj.get(node)) {
                indegree[neighbour]--;
                if (indegree[neighbour] == 0) {
                    q.offer(neighbour);
                }
            }
        }
        if (index == v) {
            return topoSort;
        }
        return new int[0];
    }

    public static void main(String[] args) {

    }
}
